File: src/NeuralNetwork/NeuralNetwork/sample.cpp
#include "sample.hpp"
#include <fstream>

Sample::Sample(dimension inputSize, dimension outputSize) : inputSize(inputSize), outputSize(outputSize) {
    input = new scalar[inputSize];
    output = new scalar[outputSize];
}

Sample::~Sample() {
    delete[] input;
    delete[] output;
}

bool Sample::save(std::string path) {
    std::ofstream out(path);
    if(out.fail()) {
        Log::print("Failed to open file");
        return false;
    }
    
    // Write 4 bytes to indicate the right file type
    out.write("NNTS", 4); // Neural Network Training Sample
    
    // Write dimension of input and output
    out.write((const char*) &inputSize, sizeof(inputSize));
    out.write((const char*) &outputSize, sizeof(outputSize));
    
    // Write data
    if(inputSize > 0) out.write((const char*) input, sizeof(scalar) * inputSize);
    if(outputSize > 0) out.write((const char*) output, sizeof(scalar) * outputSize);
    
    out.close();
    return true;
}

Sample* Sample::load(std::string path) {
    std::ifstream in(path);
    if(in.fail()) {
        Log::print("Failed to open file");
        return nullptr;
    }
    
    // Read 4 bytes
    char code[4];
    in.read(code, 4);
    
    if(std::string(code, 4).compare("NNTS") != 0) {
        Log::print("Invalid file");
        return nullptr;
    }
    
    // Read dimension of input and output
    dimension inputSize, outputSize;
    in.read((char*) &inputSize, sizeof(inputSize));
    in.read((char*) &outputSize, sizeof(outputSize));
    
    // Create new sample of given dimensions
    Sample* sample = new Sample(inputSize, outputSize);
    
    // Read data
    if(inputSize > 0) in.read((char*) sample->input, sizeof(scalar) * inputSize);
    if(outputSize > 0) in.read((char*) sample->output, sizeof(scalar) * outputSize);
    
    in.close();
    return sample;
}

Sample* Sample::duplicate() {
    // Create new sample, and copy input and output
    Sample* sample = new Sample(inputSize, outputSize);
    memcpy(sample->input, input, sizeof(scalar) * inputSize);
    memcpy(sample->output, output, sizeof(scalar) * outputSize);
    return sample;
}

File: src/python/sample.py
import struct

class Sample:

    def __init__(self, input_size, output_size):
        self.data_input = [0.0] * input_size
        self.data_output = [0.0] * output_size

    def save(self, path):
        output_file = open(path, "wb")
        output_file.write('NNTS')
        output_file.write(struct.pack('I', len(self.data_input)))
        output_file.write(struct.pack('I', len(self.data_output)))
        output_file.write(struct.pack('d' * len(self.data_input), *self.data_input))
        output_file.write(struct.pack('d' * len(self.data_output), *self.data_output))
        output_file.close()

    @staticmethod
    def load(path):
        input_file = open(path, "rb")
        code = input_file.read(4);
        if(code != 'NNTS'):
            print 'Invalid file'
            return None
        input_size = struct.unpack('I', input_file.read(4))[0];
        output_size = struct.unpack('I', input_file.read(4))[0];
        sample = Sample(input_size, output_size)
        sample.data_input = list(struct.unpack('d' * input_size, input_file.read(8 * input_size)))
        sample.data_output = list(struct.unpack('d' * output_size, input_file.read(8 * output_size)))
        return sample
